src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -169,7 +169,19 @@ impl StreamSink {
                     }
                     _ => {
                         assert_matches!(user_distributed_by, RequiredDist::Any);
-                        RequiredDist::shard_by_key(input.schema().len(), input.expect_stream_key())
+                        if downstream_pk.is_empty() {
```

Comment:
Is the sink compaction still needed if enforced shuffle?

Reply:
Yes, it happens on the shuffled data on the downstream fragment